src/models/ae.jl

```diff
@@ -0,0 +1,160 @@
+struct AE
+	encoder
+	decoder
+end
+
+Flux.@functor AE
+
+function (ae::AE)(x)
+	z = ae.encoder(x)
+	x̂ = ae.decoder(z)
+	return x̂
+end
+
+# ae_loss already in svdd
+
+function conv_ae_constructor(
+	;idim=(2,2,1), 
+	zdim::Int=1, 
+	activation="relu", 
+	kernelsizes=(1,1), 
+	channels=(1,1), 
+	scalings=(1,1),
+	init_seed=nothing, 
+	batchnorm=false, 
+	kwargs...
+)
+	# if seed is given, set it
+	(init_seed !== nothing) ? Random.seed!(init_seed) : nothing
+	# encoder 
+	encoder = Chain(
+		conv_encoder(idim, zdim, kernelsizes, channels, scalings; activation=activation, batchnorm=batchnorm)...
+	)
+	# decoder 
+	decoder = Chain(
+		conv_decoder(idim, zdim, reverse(kernelsizes), reverse(channels), reverse(scalings),
+			activation=activation, vec_output=false, vec_output_dim=nothing, batchnorm=batchnorm)...,
+			x->sigmoid.(x) # our dataset images are scaled from 0-1
+	)
+	# reset seed
+	(init_seed !== nothing) ? Random.seed!() : nothing
+
+	model = AE(encoder, decoder)
+end
+
+
+function ae_constructor(
+	;idim::Int=1, 
+	zdim::Int=1, 
+	hdim::Int=64, 
+	activation = "relu", 
+	nlayers::Int=3, 
+	init_seed=nothing, 
+	kwargs...
+	)
+	(init_seed !== nothing) ? Random.seed!(init_seed) : nothing
+
+	encoder = build_mlp(idim, hdim, zdim, nlayers, activation=activation, lastlayer="linear")
+
+	decoder = build_mlp(zdim, hdim, idim, nlayers, activation=activation, lastlayer="linear")
+
+	(init_seed !== nothing) ? Random.seed!() : nothing
+
+	return AE(encoder, decoder)
+end
+
+function anomaly_score(ae::AE, real_input; dims=(1,2,3), to_testmode::Bool=true)
+	(to_testmode == true) ? Flux.testmode!(ae) : nothing
+	score = vec(Flux.sum((ae(real_input) .- real_input).^2, dims=dims))
+	(to_testmode == true) ? Flux.testmode!(ae, false) : nothing
+	return score
+end
+
+function anomaly_score_gpu(ae::AE, real_input; batch_size=64, dims=(1,2,3), to_testmode::Bool=true)
+	real_input = Flux.Data.DataLoader(real_input, batchsize=batch_size)
+	(to_testmode == true) ? Flux.testmode!(ae) : nothing
+	ae = ae |> gpu
+	scores = Array{Float32}([])
+	for x in real_input
+		x = x |> gpu
+		score = vec(Flux.sum((ae(x) .- x).^2, dims=dims))
```

Comment:
Reuse the previous function here instead doing it all over again. When changing stuff later, you only have to make the change in one place.

Reply:
Also, you are 100% sure that the code is consistent.

Reply:
I don't understand. Consistent in what way? You mean anomaly_score and anomaly_score_gpu?

Reply:
I would define a separate one line function
```
mse_score(m,x,dims) = vec(Flux.sum((m(x) .- x).^2, dims=dims))
```
and reuse it in the two functions.